### PR TITLE
WIP: messages api

### PR DESCRIPTION
**Pull request by j-ro**
queries are awkward

not sure about the PUT method for sending/scheduling, but it might work

haven't done associated resources like email wrappers yet

### CODE REVIEW

**Comment by j-ro** [comment]
Ok, did a nice update here:

- moved scheduling and sending to helper endpoints
  - did just two endpoints -- POST creates, DELETE stops/cancels. Think this is better than making 4 endpoints.
  - thoughts on scheduling? I made POST contain a datetime, but the actual result would be seen on the parent message endpoint, where the scheduled date is shown. I think that's ok, maybe awkward though.
- moved target inputs to an array of links
  - this makes it easier than a collection to do updates -- pull the full array and PUT back your changes.
  - thoughts on the format? it's not quite a full _links section, but is an object with a href field
- did a wrapper resources
  - SMS folks, need help here adding fields for SMS wrappers, not sure what they should be. Right now it's really just working for email.

Any and all thoughts welcome! I don't think I can make this week's meeting, so can discuss here.

File: wrappers.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+A list of fields specific to the Wrapper resource.
+
+| Name          | Type      | Description
+|-----------    |-----------|-----------|--------------
+|origin_system		|string     |A human readable identifier of the system where this wrapper was created. (ex: "OSDI System")
+|name				|string		|The name of the wrapper. Intended for administrative display rather than a public title, though may be shown to a user.
+|header				|string		|The header content of the wrapper. May contain HTML.
+|footer				|string		|The footer content of the wrapper. May contain HTML.
+|administrative_url		|string		|A URL string pointing to the wrapper's administrative page on the web.
+|type				|enum		|The type of wrapper. One of "email" or "sms".
```
Comment (joshco) [inline wrappers.md]: this wrapper_type

**Comment by augfrank** [comment]
I reviewed the messages API. It looks good to me. The wrappers does not really apply to the SMS campaigns directly. SMS campaigns have a (keyword, auto-response) pair attached to them. This enables people to build keyword trees in SMS campaigns. Do you want to model this in this API? For example, there'll be a campaign that says,

From: 15104024173
Message: What is your favorite issue? Apple, Orange or Pear

Then there'll be a set of keyword auto-responses.

keyword: Apple
response: Great! I love apple too.

keyword: Orange
response: Orange is a citrus fruit

keyword: Pear
response: Did you know pear is used as an ornamental fruit?

I'm happy to talk to you over a phone call before next week's meeting as well.

**Comment by j-ro** [comment]
I think that's up to you! Do you think it's a useful thing to have over an API?

Right now, my use case for the wrappers API is to have it there to attach to messages. So our implementation won't allow POST/PUT/DELETE, just GET, and honestly the fields there aren't super important other than name and link, since that's all you really need to attach it to the message.

So, I think the question for you is, do you see a use for having all that metadata about an SMS wrapper on the API? If so, I'd say maybe you should create a fork of my branch here and add some fields and we can merge it back in?

**Comment by augfrank** [comment]
Yes, I can do that. It will be another link in this object with its own POST/PUT/DELETE. It might take a few days though. I'm okay with letting this pull request go in and we can send another with the changes for the SMS keywords.

**Comment by j-ro** [comment]
Ok great!

File: wrappers.md (hunk 1 above)
Comment (j-ro) [inline wrappers.md]: updated

**Comment by joshco** [comment]
Approved by CMTE 3/9/17

**Pull request merged**